src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,250 @@
+    * If either of the incoming reads are undefined, the duplex read will be undefined.
+    *
+    * @param ab the single-strand consensus from one strand
+    * @param ba the single-strand consensus from the other strand
+    * @return a duplex consensus if one can be built
+    */
+  private[umi] def duplexConsensus(ab: Option[VanillaConsensusRead],
+                                   ba: Option[VanillaConsensusRead],
+                                   sourceReads: Seq[SourceRead]): Option[DuplexConsensusRead] = {
+    (ab, ba) match {
+      case (Some(a), Some(b)) =>
+        val len = min(a.length, b.length)
+        val id  = a.id
+        val bases  = new Array[Byte](len)
+        val quals  = new Array[Byte](len)
+        val errors = new Array[Short](len)
+
+        forloop(from=0, until=len) { i =>
+          val aBase = a.bases(i)
+          val bBase = b.bases(i)
+          val aQual = a.quals(i)
+          val bQual = b.quals(i)
+
+          val (base, qual) = {
+            if      (aBase == NoCall || bBase == NoCall) (NoCall, PhredScore.MinValue)
+            else if (aBase == bBase) (aBase, (aQual + bQual).toByte)
+            else if (aQual > bQual)  (aBase, (aQual - bQual).toByte)
+            else if (bQual > aQual)  (bBase, (bQual - aQual).toByte)
+            else (NoCall, PhredScore.MinValue)
+          }
+
+          bases(i)  = base
+          quals(i)  = qual
+          errors(i) = sourceReads.count(s => s.bases(i) != NoCall && s.bases(i) != base).toShort
+        }
+
+        Some(DuplexConsensusRead(id=id, bases, quals, errors, a.truncate(bases.length), b.truncate(bases.length)))
+      case _ =>
+        None
+    }
+  }
+
+  /**
+    * Creates a SAMRecord with a ton of additional tags annotating the duplex read.
+    */
+  override protected def createSamRecord(read: DuplexConsensusRead, readType: ReadType): SAMRecord = {
+    val rec = super.createSamRecord(read, readType)
+    val ab = read.abConsensus
+    val ba = read.baConsensus
+
+    // Calculate the total depths across both SS consensus reads
+    val totalDepths = read.abConsensus.depths.zip(read.baConsensus.depths).map(x => x._1 + x._2)
+
+    { import ConsensusTags.PerRead._
+      rec.setAttribute(RawReadCount,       totalDepths.max)
+      rec.setAttribute(MinRawReadCount,    totalDepths.min)
+      rec.setAttribute(RawReadErrorRate,   read.errors.sum / totalDepths.sum.toFloat)
+      rec.setAttribute(AbRawReadCount,     read.abConsensus.depths.max.toInt)
+      rec.setAttribute(BaRawReadCount,     read.baConsensus.depths.max.toInt)
+      rec.setAttribute(AbMinRawReadCount,  read.abConsensus.depths.min.toInt)
+      rec.setAttribute(BaMinRawReadCount,  read.baConsensus.depths.min.toInt)
+      rec.setAttribute(AbRawReadErrorRate, read.abConsensus.errors.sum / read.abConsensus.depths.sum.toFloat)
+      rec.setAttribute(BaRawReadErrorRate, read.baConsensus.errors.sum / read.baConsensus.depths.sum.toFloat)
```

Comment:
I don't see how the sum will guard against an overflow here.  Can you add some guards against that and cap it?